networkx/algorithms/community/quality.py

```diff
@@ -276,45 +287,43 @@ def modularity(G, communities, weight="weight"):
     >>> import networkx.algorithms.community as nx_comm
     >>> G = nx.barbell_graph(3, 0)
     >>> nx_comm.modularity(G, [{0, 1, 2}, {3, 4, 5}])
-    0.35714285714285704
+    0.35714285714285715
     >>> nx_comm.modularity(G, nx_comm.label_propagation_communities(G))
-    0.35714285714285704
+    0.35714285714285715
 
     References
     ----------
     .. [1] M. E. J. Newman *Networks: An Introduction*, page 224.
        Oxford University Press, 2011.
-
+    .. [2] Clauset, Aaron, Mark EJ Newman, and Cristopher Moore.
+    "Finding community structure in very large networks."
+    Physical review E 70.6 (2004). <https://arxiv.org/abs/cond-mat/0408187>
     """
     if not isinstance(communities, list):
         communities = list(communities)
     if not is_partition(G, communities):
         raise NotAPartition(G, communities)
 
-    multigraph = G.is_multigraph()
     directed = G.is_directed()
     m = G.size(weight=weight)
-    if directed:
-        out_degree = dict(G.out_degree(weight=weight))
-        in_degree = dict(G.in_degree(weight=weight))
-        norm = 1 / m
-    else:
-        out_degree = dict(G.degree(weight=weight))
-        in_degree = out_degree
-        norm = 1 / (2 * m)
-
-    def val(u, v):
-        try:
-            if multigraph:
-                w = sum(d.get(weight, 1) for k, d in G[u][v].items())
-            else:
-                w = G[u][v].get(weight, 1)
-        except KeyError:
-            w = 0
-        # Double count self-loops if the graph is undirected.
-        if u == v and not directed:
-            w *= 2
-        return w - in_degree[u] * out_degree[v] * norm
-
-    Q = sum(val(u, v) for c in communities for u, v in product(c, repeat=2))
-    return Q * norm
+
+    def degree_sum(degree_view):
+        return sum(map(lambda node_degree: node_degree[1], degree_view))
+
+    def community_contribution(community):
+        subgraph = nx.subgraph(G, community)
+        intra_community_link_weight = subgraph.size(weight=weight)
+
+        if directed:
+            out_degree_sum = degree_sum(G.out_degree(community, weight=weight))
+            in_degree_sum = degree_sum(G.in_degree(community, weight=weight))
+            norm = 1 / m
+        else:
+            out_degree_sum = degree_sum(G.degree(community, weight=weight))
+            in_degree_sum = out_degree_sum
+            norm = 1 / (2 * m)
+
+        return intra_community_link_weight / m \
+            - (out_degree_sum * in_degree_sum * norm ** 2)
```

Comment:
This can be sped up by inlining degree_sum while at the same time, pulling out G.degree. The pulled out degree also allows quick computation of size.   Finally, forming the subgraph turns out to be slower if you only want to compute the subgraph's size.  So, I suggest something similar to the following:

```
    directed = G.is_directed() 
    if directed: 
        out_degree = dict(G.out_degree(weight=weight)) 
        in_degree = dict(G.in_degree(weight=weight)) 
        m = sum(out_degree.values()) 
        norm = 1 / m ** 2 
    else: 
        out_degree = in_degree = dict(G.degree(weight=weight)) 
        m = sum(out_degree.values()) / 2 
        norm = 1 / (2 * m) ** 2 

    def community_contribution(community): 
        comm = set(community) 
        Lc = sum(wt for u, v, wt in G.edges(comm, data=weight, default=1) if v in comm) 

        out_degree_sum = sum(out_degree[u] for u in comm) 
        in_degree_sum = sum(in_degree[u] for u in comm) if directed else out_degree_sum 

        return Lc / m - out_degree_sum * in_degree_sum * norm 
```

Reply:
Your variant definitely runs faster. I've replaced it in the code.